src/Kernel/Context.class.st

```diff
@@ -778,6 +778,17 @@ Context >> exceptionHandlerBlock [
 
 ]
 
+{ #category : #'private-exceptions' }
+Context >> exceptionsToCaptureWhenStepping [
+
+	| exceptionsToCapture exceptionsSet |
+	exceptionsToCapture := Exception allSubclasses select: [ :c | 
```

Comment:
It can decrease the performance of StepOver button in the debugger (or other related). 
And more important it affects the process termination which performs the unwinds using this simulation machinery. See the method Process>>#doTerminationFromAnotherProcess.

I think you can just implement your own ExceptionSet class:
```Smalltalk
SimulationRunTrapExceptionSet>>handles: anException
   ^anException captureIfSignalledWhenStepping

Context>>runUntilErrorOrReturnFrom: aSender
   ...
   aSender insertSender: (Context
		contextOn: SimulationRunTrapExceptionSet new do: [:ex |
   ...
```
And it all deserves better names

Reply:
And this way you will avoid the implementation like "self == Error". Simply return true or false as it really relevant for all subclasses

Reply:
> 
> `[] on: Error, Halt do: [:er | ]` is equivalent to `[] on: (ExceptionSet new add: Error; add: Halt) do: [:er | ]`
> 
> The #, message just produces an instance of ExceptionSet. You can pass there any object which understand #handles: message. Therefore you can implement your own analogue of ExceptionSet and avoid the concatenation of exception classes.

Ah, that I know, but it cannot work like that because I need to add `TestFailure` directly to the exception set as it is not in the kernel. I need to find it dynamically. 

Maybe I can build an "extendable exception set" with extensions methods, that I dynamically recover through pragmas. Looking for pragmas in a single class should be much faster than looking in the exception class hierarchy like I did but I don't know for sure. I'm ok to fix the design (it would be better with your proposal I think), but I'm not up to compare performance of both solutions. I would need help on that side.

Reply:
Hi @StevenCostiou . It seems you did not get my example:
```Smalltalk
SimulationRunTrapExceptionSet>>handles: anException
   ^anException captureIfSignalledWhenStepping

Context>>runUntilErrorOrReturnFrom: aSender
   ...
   aSender insertSender: (Context
		contextOn: SimulationRunTrapExceptionSet new do: [:ex |
   ...
```
There is no reference to TestFailure there. But it does exactly same job as your solution with classes concatenation.

Reply:
Ah. Sorry I still don't understand.
How would SimulationRunTrapExceptionSet work?
It has to:
- be in the kernel
- contain all the exception classes that interrupt the simulator
To satisfy this last point, it has to find those classes... which would have to include TestFailure.

Reply:
> Ah. Sorry I still don't understand.
> How would SimulationRunTrapExceptionSet work?
> It has to:
> 
> * be in the kernel

It's fine to be in Kernel as it represents a set of exceptions to be trapped during the simulation. Ideally all these simulation methods (runUntilErrorOrReturn and friends) should in own package but it is where we are.

> * contain all the exception classes that interrupt the simulator
>   To satisfy this last point, it has to find those classes... which would have to include TestFailure.

See the #handles: method in my example. It just asks the given exception if the simulation should be interrupted (no exception classes are hardcoded there). With my name it looks better:
```Smalltalk
SimulationRunTrapExceptionSet>>handles: anException
   ^anException shouldInterruptSimulation
```
Compare it with:
```Smalltalk
ExceptionSet>>handles: anException
	"Determine whether an exception handler will accept a signaled exception."
	^ exceptions anySatisfy: [ :exception | exception handles: anException ]
```
Does it makes sense now? I don't know how to explain it better.

Reply:
Ah, yes I understand. I missed that above in your previous message, I understand now. It greatly improves the solution.
Thank you for explaining again :)

Well I realize now... you have put the entire solution (for the simulation interruption) into this discussion... would you consider implementing it and submitting a pull request? :)

Reply:
Hi @StevenCostiou .

I did not know when I will have time for this. Feel free to push it forward